apps/app/src/views/routes/Settings/components/users/EditRolesAction.tsx

```diff
@@ -0,0 +1,444 @@
+            postalCode: yup
+              .string()
+              .required('Zip is required')
+              .matches(/^[0-9]{5}(?:-[0-9]{4})?$/, { message: 'Enter a valid zipcode' })
+          })
+          .required('Please enter an address'),
+        phone: yup
+          .string()
+          .matches(
+            /^\s*(?:\+?(\d{1,3}))?[-. (]*(\d{3})[-. )]*(\d{3})[-. ]*(\d{4})(?: *x(\d+))?\s*$/,
+            { message: 'Please enter a valid phone number' }
+          )
+          .required('Please enter a valid phone number')
+      })
+      .notRequired()
+      .default(undefined)
+      .when('roles', (roles: { value: string; label: string }[], schema) => {
+        return hasPrescriberRole(roles)
+          ? schema.required(
+              "Address, NPI and phone are required for users with 'Prescriber' permissions"
+            )
+          : schema.notRequired();
+      })
+  })
+  // If not a prescriber, set the provider to undefined
+  .transform((value) => {
+    if (hasPrescriberRole(value.roles)) {
+      return value;
+    }
+    return { ...value, provider: undefined };
+  });
+
+type RoleYupType = yup.InferType<typeof roleSchema>;
+type ProviderYupType = RoleYupType['provider'];
+type ProviderFormikTouchedType = FormikTouched<ProviderYupType>;
+type ProviderFormikErrorsType = FormikErrors<ProviderYupType>;
+
+export const EditRolesAction: React.FC<EditRolesActionProps> = ({ userId, onClose }) => {
+  const client = useClinicalApiClient();
+  const {
+    data: userData,
+    error: userDataError,
+    loading
+  } = useQuery(GetUserQuery, {
+    client,
+    variables: { userId: userId }
+  });
+  const [setUserRoles, { error }] = useMutation(SetUserRolesMutation, {
+    client,
+    refetchQueries: ['UsersListQuery', 'usersQuery']
+  });
+  const [updateProviderProfile] = useMutation(UpdateProviderProfileMutation, {
+    client,
+    refetchQueries: []
+  });
+
+  const handleSaveRoles = (formVariables: any) => {
+    if (
+      hasPrescriberRole(formVariables.variables.roles) &&
+      (userData?.user?.npi == undefined ||
+        userData?.user?.address == undefined ||
+        userData?.user?.email == undefined)
+    ) {
+      updateProviderProfile({
+        variables: {
+          providerId: userId,
+          updateProviderProfileInput: {
+            address: formVariables.variables.provider.address ?? userData?.user?.address,
+            email: formVariables.variables.email ?? userData?.user?.email,
+            npi: formVariables.variables.provider.npi ?? userData?.user?.npi,
+            phone: formVariables.variables.phone ?? userData?.user?.phone
+          }
+        }
+      });
+    }
+    setUserRoles({
+      variables: {
+        userId: userId,
+        roles: formVariables.variables.roles.map((role: any) => role.value)
+      }
+    });
+    onClose;
+  };
+
+  const initialValues: yup.InferType<typeof roleSchema> = {
+    email: userData?.user?.email ?? '',
+    roles: [],
```

Comment:
this should start with the user's existing roles

Reply:
done